sklearn/pipeline.py

```diff
@@ -177,20 +177,31 @@ def _validate_steps(self):
         for t in transformers:
             if t is None or t == 'passthrough':
                 continue
-            if (not (hasattr(t, "fit") or hasattr(t, "fit_transform")) or not
-                    hasattr(t, "transform")):
+            if (not hasattr(t, "fit_transform") or
+                    (not (hasattr(t, "fit") and hasattr(t, "transform")))):
```

Comment:
It should be an `and` at the end of line 180, no?
```python
if not fit_transform AND not(fit and transform)
```
If it doesn't have a `fit_transform` method, the condition will be True currently. That's why some tests are failing.

Reply:
Yeah, that makes sense.

Reply:
Also fixed it for `FeatureUnion._validate_transformers()`.